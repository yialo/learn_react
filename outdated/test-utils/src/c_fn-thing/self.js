import * as React from 'react';

export function SubThing() {
  return <p className="sub">Sub</p>;
}

export function Thing() {
  return (
    <div>
      <SubThing key="cool" foo="bar" />
      <p className="my">Hello</p>
    </div>
  );
}
