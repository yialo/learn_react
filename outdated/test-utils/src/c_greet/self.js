export const Greet = ({ name }) => {
  if (name) {
    return <h2>Hello, {name}</h2>;
  }

  return <p>Hey, stranger</p>;
};
