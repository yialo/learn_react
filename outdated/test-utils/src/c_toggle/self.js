import * as React from 'react';

export function Toggle(props) {
  const [isOn, setIsOn] = React.useState(false);

  return (
    <button
      onClick={() => {
        setIsOn((previousState) => !previousState);
        props.onChange(!isOn);
      }}
      data-testid="toggle"
    >
      {isOn ? 'Turn on' : 'Turn off'}
    </button>
  );
}
