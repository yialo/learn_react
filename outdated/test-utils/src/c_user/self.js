import * as React from 'react';

export const User = ({ id }) => {
  const [user, setUser] = React.useState(null);

  React.useEffect(() => {
    const fetchUser = async () => {
      const response = await fetch(`/${id}`);
      setUser(await response.json());
    };

    fetchUser();
  }, [id]);

  if (!user) {
    return <div>Loading...</div>;
  }

  return (
    <details>
      <summary>{user.name}</summary>
      <p>
        {'Age: '}
        <strong>{user.age}</strong>
        {' years old'}
      </p>
      <p>{`Lives in ${user.address}`}</p>
    </details>
  );
};
