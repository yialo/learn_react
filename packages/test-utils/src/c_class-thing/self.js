import * as React from 'react';

export class Thing extends React.Component {
  constructor(props) {
    super(props);
    this.input = null;
  }
  componentDidMount() {
    this.input.focus();
  }
  render() {
    return <input type="text" ref={(el) => (this.input = el)} />;
  }
}
