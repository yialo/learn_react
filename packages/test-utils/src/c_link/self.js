import * as React from 'react';

const CLASSNAMES = {
  NORMAL: 'NORMAL',
  HOVERED: 'HOVERED',
};

export const Link = ({ children, href }) => {
  const [className, setClassName] = React.useState(CLASSNAMES.NORMAL);

  return (
    <a
      className={className}
      href={href ?? '#'}
      onMouseEnter={() => {
        setClassName(CLASSNAMES.HOVERED);
      }}
      onMouseLeave={() => {
        setClassName(CLASSNAMES.NORMAL);
      }}
    >
      {children}
    </a>
  );
};
