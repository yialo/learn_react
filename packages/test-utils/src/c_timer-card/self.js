import * as React from 'react';

const CHOICES = Array.from({ length: 4 }).map((_, i) => i + 1);
const DELAY = 5000;

export const TimerCard = ({ onSelect }) => {
  const ref_lastOnSelect = React.useRef(onSelect);

  React.useEffect(() => {
    ref_lastOnSelect.current = onSelect;
  }, [onSelect]);

  React.useEffect(() => {
    const timerId = setTimeout(() => {
      ref_lastOnSelect.current(null);
    }, DELAY);

    return () => {
      clearTimeout(timerId);
    };
  }, []);

  return CHOICES.map((choice) => {
    return (
      <button
        key={choice}
        data-testid={choice}
        onClick={() => {
          onSelect(choice);
        }}
      >
        {choice}
      </button>
    );
  });
};
