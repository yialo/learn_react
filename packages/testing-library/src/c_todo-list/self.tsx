import * as React from 'react';

type TodoEntity = {
  id: number;
  title: string;
  isComplete: boolean;
};

type State = {
  todos: TodoEntity[];
  isLoading: boolean;
  error: Error | null;
};

const INITIAL_STATE: State = {
  todos: [],
  isLoading: true,
  error: null,
};

const TODOS: TodoEntity[] = [
  {
    id: 1,
    title: 'Item 1',
    isComplete: false,
  },
  {
    id: 2,
    title: 'Item 2',
    isComplete: false,
  },
  {
    id: 3,
    title: 'Item 3',
    isComplete: false,
  },
];

export const TodoList = () => {
  const [state, setState] = React.useState(INITIAL_STATE);

  React.useEffect(() => {
    setTimeout(() => {
      setState((prev) => ({
        ...prev,
        isLoading: false,
        todos: TODOS,
      }));
    }, 500);
  }, []);

  const isLoaded = !state.isLoading && !state.error;
  const hasErorr = !state.isLoading && state.error;

  return (
    <div>
      {state.isLoading && <div>Loading...</div>}

      {hasErorr && <div role="alert">Loading failed</div>}

      {isLoaded &&
        (state.todos.length > 0 ? (
          <>
            <p>My todos:</p>
            <ul>
              {state.todos.map(({ id, isComplete, title }) => {
                const handleClick = () => {
                  setState((prev) => {
                    const targetTodo = prev.todos.find((todo) => todo.id === id);

                    if (!targetTodo) {
                      return prev;
                    }

                    const newTodos = prev.todos.map((todo) => {
                      if (todo !== targetTodo) {
                        return todo;
                      }

                      return {
                        ...todo,
                        isComplete: !todo.isComplete,
                      };
                    });

                    return {
                      ...prev,
                      todos: newTodos,
                    };
                  });
                };

                return (
                  <li key={id}>
                    <div>{title}</div>
                    <div>{`Completed: ${isComplete}`}</div>
                    <button type="button" aria-pressed={isComplete} onClick={handleClick}>
                      Check
                    </button>
                  </li>
                );
              })}
            </ul>
          </>
        ) : (
          <div>No todos</div>
        ))}
    </div>
  );
};
