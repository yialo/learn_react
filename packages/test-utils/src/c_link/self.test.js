import renderer from 'react-test-renderer';

import { Link } from './self';

describe('Link', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Link href="https://github.com/">GitHub</Link>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders an anchor when no href is passed', () => {
    const tree = renderer.create(<Link>GitHub</Link>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('updates CSS class on hover', () => {
    const component = renderer.create(<Link href="https://github.com/">GitHub</Link>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
      tree.props.onMouseEnter();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
      tree.props.onMouseLeave();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
