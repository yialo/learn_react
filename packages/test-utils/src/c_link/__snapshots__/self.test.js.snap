// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders an anchor when no href is passed 1`] = `
<a
  className="NORMAL"
  href="#"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  GitHub
</a>
`;

exports[`Link renders correctly 1`] = `
<a
  className="NORMAL"
  href="https://github.com/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  GitHub
</a>
`;

exports[`Link updates CSS class on hover 1`] = `
<a
  className="NORMAL"
  href="https://github.com/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  GitHub
</a>
`;

exports[`Link updates CSS class on hover 2`] = `
<a
  className="HOVERED"
  href="https://github.com/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  GitHub
</a>
`;

exports[`Link updates CSS class on hover 3`] = `
<a
  className="NORMAL"
  href="https://github.com/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  GitHub
</a>
`;
